network/p2p/gossip/mempool.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+
+	xmempool "github.com/ava-labs/avalanchego/vms/txs/mempool"
+)
+
+var ErrDuplicateTx = xmempool.ErrDuplicateTx
+
+type Mempool[T Gossipable] interface {
+	Add(tx T) error
+	Remove(txs ...T)
+
+	// Len returns the number of txs in the mempool.
+	Len() int
+
+	// Has returns true if the gossipable is in the set.
+	Has(gossipID ids.ID) bool
+}
+
+type mempool[T Gossipable] struct {
+	Txs     map[ids.ID]bool
+	metrics *Metrics
+}
+
+func NewMempool[T Gossipable](
+	metrics *Metrics,
+) (*mempool[T], error) {
+	m := &mempool[T]{
+		Txs:     make(map[ids.ID]bool),
+		metrics: metrics,
+	}
+	return m, nil
+}
+
+func (m *mempool[T]) Add(tx T) error {
+	txID := tx.GossipID()
+	if m.Has(txID) {
+		m.metrics.ObserveIncomingGossipable(txID, DroppedDuplicate)
+		return fmt.Errorf("%w: %s", ErrDuplicateTx, txID)
```

Comment:
We only return an error in the duplicate case... so do we even need to return an error if it's always the same case? I think we could just return a true/false bool here

Reply:
If we're going to use this struct to be a replacement for the real mempool, I'd say "yes, let's keep it".
Keep in mind that it also save the `gossipMempool` in platformvm and avm from having the need to format that message.